libs/ui-components/src/components/modals/ApproveDeviceModal/ApproveDeviceForm.tsx

```diff
@@ -0,0 +1,173 @@
+  error?: string;
+  children?: React.ReactNode;
+};
+
+type DeviceMatchStatus = {
+  status: 'unchecked' | 'checking' | 'checked--unique' | 'checked--multiple' | 'checked--empty';
+  fleetName?: string;
+};
+
+const findMatchingFleets = (fleets: Fleet[], deviceLabels: FlightCtlLabel[]) => {
+  return fleets.filter((fleet) => {
+    const fleetMatch = fleet.spec.selector?.matchLabels || {};
+    return Object.entries(fleetMatch).every(([fleetMatchKey, fleetMatchValue]) => {
+      const matchingDeviceLabel = deviceLabels.find((dLabel) => dLabel.key === fleetMatchKey);
+      return matchingDeviceLabel && (matchingDeviceLabel.value || '') === (fleetMatchValue || '');
+    });
+  });
+};
+
+const FleetLabelMatchResult = ({ matchStatus }: { matchStatus: DeviceMatchStatus }) => {
+  const { t } = useTranslation();
+
+  let icon: React.ReactNode;
+  let text: string = '';
+  let isPlaceholder = false;
+
+  switch (matchStatus.status) {
+    case 'unchecked':
+      isPlaceholder = true;
+      text = t('Add labels to select a fleet');
+      break;
+    case 'checking':
+      icon = <Spinner size="sm" />;
+      break;
+    case 'checked--empty':
+      icon = icon = (
+        <Icon status="warning">
+          <ExclamationTriangleIcon />
+        </Icon>
+      );
+      text = t('No fleet is matching the selected labels');
+      break;
+    case 'checked--multiple':
+      icon = (
+        <Icon status="danger">
+          <ExclamationCircleIcon />
+        </Icon>
+      );
+      text = t(
+        "More than one fleet is matching the selected labels. The device will ignore the fleets' configurations",
+      );
+      break;
+    case 'checked--unique':
+      text = matchStatus.fleetName || '';
+      break;
+  }
+
+  return (
+    <div className="fctl_approve-device__fleetname">
+      {icon && <span className="fctl_approve-device__fleetname__icon">{icon}</span>}
+      <span className={isPlaceholder ? 'fctl_approve-device__fleetname__placeholder' : ''}>{text}</span>
+    </div>
+  );
+};
+
+const ApproveDeviceForm: React.FC<ApproveDeviceFormProps> = ({ enrollmentRequest, onClose, error, children }) => {
+  const { t } = useTranslation();
+  const { get } = useFetch();
+  const { submitForm, isSubmitting, errors: formErrors } = useFormikContext<ApproveDeviceFormValues>();
+  const [matchStatus, setMatchStatus] = React.useState<DeviceMatchStatus>({
+    status: 'unchecked',
+    fleetName: '',
+  });
+
+  const disableSubmit = Object.keys(formErrors).length > 0;
+  const isLoading = isSubmitting || matchStatus.status === 'checking';
+
+  const onChangeLabels = React.useCallback(
+    (newLabels: FlightCtlLabel[]) => {
+      const updateMatchStatus = async () => {
+        if (newLabels.length > 0) {
+          setMatchStatus({ status: 'checking' });
+
+          // TODO PoC implementation, we're missing the ability of filtering in the "fleets" endpoint.
+          const allFleets = await get<FleetList>('fleets');
```

Comment:
missing error handling

Reply:
also we may have an issue with concurrency here. We should cancel the previous request if it is running.

Reply:
Added error handling and the abortController to the new hook.